src/server/api/user_api.py

```diff
@@ -0,0 +1,308 @@
+            """select password, pdp_user_roles.role, active 
+                from pdp_users 
+                left join pdp_user_roles on pdp_users.role = pdp_user_roles._id 
+                where username=:u """
+        )
+        s = s.bindparams(u=request.form["username"])
+        result = connection.execute(s)
+
+        if result.rowcount:  # Did we get a match on username?
+            pwhash, role, is_active = result.fetchone()
+        else:
+            log_user_action(request.form["username"], "Failure", "Invalid username")
+            return jsonify("Bad credentials"), 401
+
+        if is_active.lower() == "y" and check_password(
+            request.form["password"], pwhash
+        ):
+            # Yes, user is active and password matches
+            token = jwt_ops.create_token(request.form["username"], role)
+            log_user_action(request.form["username"], "Success", "Logged in")
+            return token
+
+        else:
+            log_user_action(
+                request.form["username"], "Failure", "Bad password or inactive"
+            )
+            return jsonify("Bad credentials"), 401
+
+
+# JSON login version
+@user_api.route("/api/user/login_json", methods=["POST"])
+def user_login_json():
+    """ Validate user in db, return JWT if legit and active.
+        Expects json-encoded form data
+    """
+
+    post_dict = json.loads(request.data)
+    username = post_dict["username"]
+    presentedpw = post_dict["password"]
+
+    with engine.connect() as connection:
+
+        pwhash = None
+        s = text(
+            """select password, pdp_user_roles.role, active 
+                from pdp_users 
+                left join pdp_user_roles on pdp_users.role = pdp_user_roles._id 
+                where username=:u """
+        )
+        s = s.bindparams(u=username)
+        result = connection.execute(s)
+
+        if result.rowcount:  # Did we get a match on username?
+            pwhash, role, is_active = result.fetchone()
+        else:
+            log_user_action(username, "Failure", "Invalid username")
+            return jsonify("Bad credentials"), 401
+
+        if is_active.lower() == "y" and check_password(presentedpw, pwhash):
+            # Yes, user is active and password matches
+            token = jwt_ops.create_token(username, role)
+            log_user_action(username, "Success", "Logged in")
+            return token
+
+        else:
+            log_user_action(username, "Failure", "Bad password or inactive")
+            return jsonify("Bad credentials"), 401
+
+
+# Logout is not strictly needed; client can just delete JWT, but good for logging
+@user_api.route("/api/user/logout", methods=["POST"])
+def user_logout():
+    # Lookup user in db
+    username = request.form["username"]
+
+    # For now, just echo the data
+    log_user_action(username, "Success", "Logged out ")
+    return jsonify("Logged out " + username)
+
+
+@user_api.route("/api/user/create", methods=["POST"])
```

Comment:
maybe this endpoint should move to admin_api?

Reply:
Sure. It's protected by the JWT admin_required.